framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -1063,6 +1064,22 @@ public void postProcessClassTree(ClassTree tree) {
             // do so is here.
             wholeProgramInference.saveResults();
         }
+
+        if (tree.getExtendsClause() == null) {
+            return;
+        }
+
+        // if "@B class Y extends @A X {}", @B must be a subtype of @A
+        Set<? extends AnnotationMirror> topAnnotations = qualHierarchy.getTopAnnotations();
+        for (AnnotationMirror topAnno : topAnnotations) {
+            AnnotationMirror classType = getAnnotatedType(tree).getAnnotationInHierarchy(topAnno);
+            AnnotationMirror extendsType =
+                    getAnnotatedType(tree.getExtendsClause()).getAnnotationInHierarchy(topAnno);
+            if (!qualHierarchy.isSubtype(classType, extendsType)) {
+                @CompilerMessageKey String errorKey = "extends.clause.invalid";
+                checker.report(Result.failure(errorKey, classType, extendsType), tree);
```

Comment:
Validation rules should be enforced in the Visitor https://github.com/typetools/checker-framework/blob/master/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java, not in the AnnotatedTypeFactory.
The ATF would add default/implicits for this location. I would expect that similarly to #2266 we will need a new type use location to determine what should apply to this location.

Reply:
Thanks @wmdietl I moved the rules to BaseTypeVisitor.java.